electron/Board.js

```diff
@@ -38,42 +31,120 @@ class Board {
 
   sendPacket(id, values) {
     const p = new Packet(id, values);
-    this.port.send(this.address, p.stringify());
+    const buf = p.toBuffer()
+    if(!buf){
+      console.debug(`[id ${id}] Outbound packet's packet data typing definition could not be found. Not sent.`)
+      return false
+    }
+    this.port.send(this.address, buf);
     return true;
   }
 
+  /**
+   * Uses the offset given in the first packet received from the board to calculate subsequent packet arrival times
+   * @param runTime {number} is the received running duration of the board (in ms)
+   * @returns {number} the estimated timestamp at which the packet was sent (in ms)
+   */
+  calculateTimestamp(runTime) {
+    if (this.firstRecvTime < 0) {
+      /* TODO: consider using multiple packet offsets to reduce likelihood of noise causing the first receive time to
+      *   deviate too significantly */
+      this.firstRecvTime = Date.now()
+    }
+    if (this.firstRecvOffset < 0) {
+      this.firstRecvOffset = runTime
+    }
+    return runTime - this.firstRecvOffset + this.firstRecvTime
+  }
+
+  /**
+   * Parses the raw packet into a data object
+   * @param buf {Buffer} is the buffer that contains the full udp packet content
+   * @returns {Packet|null} packet with parsed data
+   */
+  parseMsgBuf(buf) {
+    // Packet format:
+    // [ ________ | ________ | ________ ________ ________ ________ | ________ ________ | ________ ... ________ ]
+    // [    id    |   len    |              runTime                |       checkSum    |          data         ]
+    // [  u_int8  |  u_int8  |              u_int32                |        u_int16    |     defined in doc    ]
+    const id = buf.readUInt8(0);
+    const len = buf.readUInt8(1);
+
+    const runTime = buf.readUInt32LE(2);
+    const timestamp = this.calculateTimestamp(runTime);
+
+    const checksum = buf.readUInt16LE(6);
+
+    // currently, data comes after the 2 bytes checksum (at offset 2) 2 + 2 = 4
+    const dataOffset = 8;
+
+    const dataBuf = buf.slice(dataOffset, dataOffset + len)
+    const payloadBuf = buf.slice(0, 6)
+    const sumBuf = Buffer.concat([payloadBuf, dataBuf])
+    const expectedChecksum = Packet.fletcher16(sumBuf)
+
+    if (checksum === expectedChecksum) {
+      const values = []
+      const packetDef = INBOUND_PACKET_DEFS[id]
```

Comment:
Right after this line, can you add something like `if(!packetDef) return null` to cover the case where that packet ID doesn't exist in the definition